### PR TITLE
Ryna-LB109.ag_an_ALL_R_C_Y

### PR DESCRIPTION
**Pull request by rynacui**
#270

### CODE REVIEW

File: R/zchunk_LB109.ag_an_ALL_R_C_Y.R [hunk 1]
```diff
@@ -33,61 +33,209 @@ module_aglu_LB109.ag_an_ALL_R_C_Y_DISABLED <- function(command, ...) {
...
+      mutate(year = as.integer(substr(year, 2, 5)))   # change Xyear to year
+
+    L122.in_Mt_R_C_Yh <- get_data(all_data, "temp-data-inject/L122.in_Mt_R_C_Yh") %>%
+      # The following two lines of code will be removed later, when we're using 'real' data
+      gather(year, value, -GCAM_region_ID, -GCAM_commodity) %>%   # reshape
+      mutate(year = as.integer(substr(year, 2, 5)))   # change Xyear to year
+
+    # Part 1: Primary agricultural goods
+    # List of all flows for primary agricultural good balances
+    ag_Flow_cols <- c( "Prod_Mt", "NetExp_Mt", "Supply_Mt", "Food_Mt", "Feed_Mt", "Biofuels_Mt", "OtherUses_Mt" )
```
Comment (bpbond) [inline R/zchunk_LB109.ag_an_ALL_R_C_Y.R]: No GCAM-style spacing with parentheses and brackets please

File: R/zchunk_LB109.ag_an_ALL_R_C_Y.R [hunk 2]
```diff
@@ -33,61 +33,209 @@ module_aglu_LB109.ag_an_ALL_R_C_Y_DISABLED <- function(command, ...) {
...
+    # List of any commodities (e.g. pasture, residue, scavenging) in feed but not in production tables
+    L103.ag_Prod_Mt_R_C_Y %>%
+      select(GCAM_commodity) %>%
+      unique() %>%
+      .[['GCAM_commodity']] -> Primary_commodities
```
Comment (bpbond) [inline R/zchunk_LB109.ag_an_ALL_R_C_Y.R]: Double quotes please, not single.

File: R/zchunk_LB109.ag_an_ALL_R_C_Y.R [hunk 3]
```diff
@@ -33,61 +33,209 @@ module_aglu_LB109.ag_an_ALL_R_C_Y_DISABLED <- function(command, ...) {
...
+      L109.an_ALL_Mt_R_C_Y %>%
+        # Subset negative and positive "other uses" separately
+        mutate(NegOtherUses_Mt = if_else(OtherUses_Mt < 0, OtherUses_Mt, 0),
+               # Positive will be the new adjusted "other uses", and replace negative with zero
+               OtherUses_Mt_adj = if_else(OtherUses_Mt >= 0, OtherUses_Mt, 0),
+               # Assign negative "other uses" to imports, and calculate the adjusted regional net exports
+               NetExp_Mt_adj = NetExp_Mt + NegOtherUses_Mt) %>%
+        group_by(GCAM_commodity, year) %>%
+        # Calculate the changes in global net exports = sum of negative other uses, global other uses =  sum of positive other uses
+        summarise_at(vars(GlobalNetExpAdj = NegOtherUses_Mt, GlobalOtherUses_Mt = OtherUses_Mt_adj), sum) -> L109.an_ALL_Mt_glbl_C_Y
```
Comment (bpbond) [inline R/zchunk_LB109.ag_an_ALL_R_C_Y.R]: Here and elsewhere, would it be clearer to say 
```R
  summarise(GlobalNetExpAdj = sum(NegOtherUses_Mt), 
            GlobalOtherUses_Mt = sum(OtherUses_Mt_adj))
```

I don't have a strong preference but consider.

File: R/zchunk_LB109.ag_an_ALL_R_C_Y.R [hunk 4]
```diff
@@ -33,61 +33,209 @@ module_aglu_LB109.ag_an_ALL_R_C_Y_DISABLED <- function(command, ...) {
...
+
+    L122.in_Mt_R_C_Yh <- get_data(all_data, "temp-data-inject/L122.in_Mt_R_C_Yh") %>%
+      # The following two lines of code will be removed later, when we're using 'real' data
+      gather(year, value, -GCAM_region_ID, -GCAM_commodity) %>%   # reshape
+      mutate(year = as.integer(substr(year, 2, 5)))   # change Xyear to year
+
+    # Part 1: Primary agricultural goods
+    # List of all flows for primary agricultural good balances
+    ag_Flow_cols <- c( "Prod_Mt", "NetExp_Mt", "Supply_Mt", "Food_Mt", "Feed_Mt", "Biofuels_Mt", "OtherUses_Mt" )
+    # List of any commodities (e.g. pasture, residue, scavenging) in feed but not in production tables
```
Comment (kvcalvin) [inline R/zchunk_LB109.ag_an_ALL_R_C_Y.R]: This comment seems wrong. Isn't this code getting the names of the commodities that are in the production table?

File: R/zchunk_LB109.ag_an_ALL_R_C_Y.R [hunk 5]
```diff
@@ -33,61 +33,209 @@ module_aglu_LB109.ag_an_ALL_R_C_Y_DISABLED <- function(command, ...) {
...
+    L103.ag_Prod_Mt_R_C_Y %>%
+      select(GCAM_commodity) %>%
+      unique() %>%
+      .[['GCAM_commodity']] -> Primary_commodities
+    L108.ag_Feed_Mt_R_C_Y %>%
+      filter(!(GCAM_commodity %in% Primary_commodities)) %>%
+      select(GCAM_commodity) %>%
+      unique() %>%
+      .[['GCAM_commodity']] -> Feed_commodities
```
Comment (kvcalvin) [inline R/zchunk_LB109.ag_an_ALL_R_C_Y.R]: Oh, now I see. This is where we find feed only commodities. Can you add more comments so this is obvious sooner?

File: R/zchunk_LB109.ag_an_ALL_R_C_Y.R [hunk 6]
```diff
@@ -33,61 +33,209 @@ module_aglu_LB109.ag_an_ALL_R_C_Y_DISABLED <- function(command, ...) {
...
+      mutate(flow = "NetExp_Mt") %>%
+      bind_rows(mutate(L103.ag_Prod_Mt_R_C_Y, flow = "Prod_Mt")) %>%
+      bind_rows(mutate(L101.ag_Food_Mt_R_C_Y, flow = "Food_Mt")) %>%
+      bind_rows(mutate(L108.ag_Feed_Mt_R_C_Y, flow = "Feed_Mt")) %>%
+      bind_rows(mutate(L122.in_Mt_R_C_Yh, flow = "Biofuels_Mt")) %>%
+      # Get all combinations of each GCAM_commodity and flow, by spreading to wide format
+      spread(flow, value) %>%
+      # Set missing values in the complete combinations to zero
+      mutate_if(is.numeric, funs(replace(., is.na(.), 0))) %>%
+      # For any feed commodities (e.g. pasture, residue, scavenging), set production = feed
```
Comment (kvcalvin) [inline R/zchunk_LB109.ag_an_ALL_R_C_Y.R]: Why?

File: R/zchunk_LB109.ag_an_ALL_R_C_Y.R [hunk 7]
```diff
@@ -33,61 +33,209 @@ module_aglu_LB109.ag_an_ALL_R_C_Y_DISABLED <- function(command, ...) {
...
+      L109.ag_ALL_Mt_R_C_Y %>%
+        # Subset negative and positive "other uses" separately
+        mutate(NegOtherUses_Mt = if_else(OtherUses_Mt < 0, OtherUses_Mt, 0),
+               # Positive will be the new adjusted "other uses", and replace negative with zero
+               OtherUses_Mt_adj = if_else(OtherUses_Mt >= 0, OtherUses_Mt, 0),
+               # Assign negative "other uses" to imports, and calculate the adjusted regional net exports
+               NetExp_Mt_adj = NetExp_Mt + NegOtherUses_Mt) %>%
+        group_by(GCAM_commodity, year) %>%
+        # Calculate the changes in global net exports = sum of negative other uses, global other uses =  sum of positive other uses
+        summarise_at(vars(GlobalNetExpAdj = NegOtherUses_Mt, GlobalOtherUses_Mt = OtherUses_Mt_adj), sum) -> L109.ag_ALL_Mt_glbl_C_Y
```
Comment (kvcalvin) [inline R/zchunk_LB109.ag_an_ALL_R_C_Y.R]: I kind of agree with Ben...this is easier to read as `summarize(GlobalNetExpAdj = sum(NegOtherUses_Mt, etc. )`

**Comment by bpbond** [comment]
@rynacui this is good to merge I think? You've addressed all of Kate's comments too?

**Comment by rynacui** [comment]
Yes. I've addressed all comments.

**Pull request merged**